src/database-postgresql.cpp

```diff
@@ -0,0 +1,319 @@
+	infostream << std::endl;
+}
+
+
+void Database_PostgreSQL::dbconnect()
+{
+	m_conn = PQconnectdb(m_conninfo.c_str());
+
+	if (PQstatus(m_conn) != CONNECTION_OK) {
+		throw FileNotGoodException(std::string(
+			"PostgreSQL database error: ") +
+			PQerrorMessage(m_conn));
+	}
+
+	m_pgVersion = PQserverVersion(m_conn);
+
+	if (m_pgVersion < 90300) {
+		throw FileNotGoodException("PostgreSQL database error: "
+			"Server version 9.3 or greater required.");
+	}
+
+	createDatabase();
+
+	infostream << "PostgreSQL datbase: Server version " << m_pgVersion;
+	infostream << std::endl;
+	infostream << "PostgreSQL: Database connection made." << std::endl;
+}
+
+
+void Database_PostgreSQL::verifyDatabase()
+{
+	if (PQstatus(m_conn) == CONNECTION_OK)
+		return;
+
+	PQreset(m_conn);
+	ping();
+}
+
+
+bool Database_PostgreSQL::ping()
+{
+	if (PQping(m_conninfo.c_str()) != PQPING_OK) {
+		throw FileNotGoodException(std::string(
+			"PostgreSQL database error: ") +
+			PQerrorMessage(m_conn));
+	}
+	return true;
+}
+
+
+bool Database_PostgreSQL::initialized()
+{
+	return (PQstatus(m_conn) == CONNECTION_OK);
+}
+
+
+void Database_PostgreSQL::prepare()
+{
+	if (m_pgVersion < 90500) {
+		prepareStatement("save",
+			"INSERT INTO blocks (x, y, z, data) VALUES ($1::int, "
+			"$2::int, $3::int, $4::bytea)");
+		prepareStatement("update",
+			"UPDATE blocks SET data = $4::bytea WHERE x = $1::int "
+			"AND y = $2::int AND z = $3::int");
+	} else {
+		prepareStatement("save", "INSERT INTO blocks (x, y, z, data) "
+			"VALUES($1::int, $2::int, $3::int, $4::bytea) ON "
+			"CONFLICT UPDATE SET data = $4::bytea;");
+	}
+	prepareStatement("load",
+		"SELECT data FROM blocks WHERE x = $1::int AND y = $2::int AND "
+		"z = $3::int");
+	prepareStatement("delete",
+		"DELETE FROM blocks WHERE x = $1::int AND y = $2::int AND "
+		"z = $3::int");
+
+	infostream << "Postgres: SQL statements prepared." << std::endl;
+}
+
+
+void Database_PostgreSQL::prepareStatement(std::string name, std::string sql)
+{
+	resultsCheck(PQprepare(m_conn, name.c_str(), sql.c_str(), 0, NULL));
+}
+
+
+bool Database_PostgreSQL::resultsCheck(PGresult* res, bool clear /*= true*/,
+	ExecStatusType expected /*= PGRES_COMMAND_OK*/)
+{
+	if (PQresultStatus(res) != expected) {
+		throw FileNotGoodException(std::string(
+			"PostgreSQL database error: ")
+			+ (res ? PQresultErrorMessage(res) :
+			"Result is NULL!"));
+	}
+
+	if (clear)
+		PQclear(res);
+	else
+		m_res = res;
+
+	return true;
+}
+
+
+bool Database_PostgreSQL::execPrepared(std::string name, int nParams,
+	const v3s16 &pos, bool clear /*= true*/,
+	const std::string &data /*= ""*/,
+	ExecStatusType expected /*= PGRES_COMMAND_OK*/)
+{
+	const char *values[nParams];
+	int         lengths[nParams];
+	int         formats[nParams];
+
+	values[0] = itos(pos.X).c_str();
+	values[1] = itos(pos.Y).c_str();
+	values[2] = itos(pos.Z).c_str();
+
+	lengths[0] = sizeof(values[0]);
+	lengths[1] = sizeof(values[1]);
+	lengths[2] = sizeof(values[2]);
```

Comment:
This will work because sizeof is computed at runtime for VLAs. However, values[0] is a pointer to char so I'm not entirely sure what this code is doing...